compiler/src/language_server/symbol.re

```diff
@@ -0,0 +1,246 @@
+  /**
+   * Represents programming constructs like variables, classes, interfaces etc.
+   * that appear in a document. Document symbols can be hierarchical and they
+   * have two ranges: one that encloses its definition and one that points to its
+   * most interesting range, e.g. the range of an identifier.
+   */
+  [@deriving yojson]
+  type document_symbol = {
+    /**
+     * The name of this symbol. Will be displayed in the user interface and
+     * therefore must not be an empty string or a string only consisting of
+     * white spaces.
+     */
+    name: string,
+    /**
+     * More detail for this symbol, e.g the signature of a function.
+     */
+    [@default None]
+    detail: option(string),
+    /** The kind of this symbol. */
+    kind: symbol_kind,
+    /** Tags for this document symbol. */
+    [@default None]
+    tags: option(list(symbol_tag)),
+    /**
+     * The range enclosing this symbol not including leading/trailing whitespace
+     * but everything else like comments. This information is typically used to
+     * determine if the clients cursor is inside the symbol to reveal it  in the
+     * UI.
+     */
+    range: Protocol.range,
+    /**
+     * The range that should be selected and revealed when this symbol is being
+     * picked, e.g. the name of a function. Must be contained by the `range`.
+     */
+    [@key "selectionRange"]
+    selection_range: Protocol.range,
+    /**
+     * Children of this symbol, e.g. properties of a class.
+     */
+    [@default None]
+    children: option(list(document_symbol)),
+  };
+  [@deriving yojson]
+  type t = list(document_symbol);
+};
+
+// Outline building functions
+let rec build_toplevel_outline = (statement: Typedtree.toplevel_stmt) => {
+  switch (statement.ttop_desc) {
+  | TTopForeign(_)
+  | TTopInclude(_)
+  | TTopProvide(_)
+  | TTopException(_)
+  | TTopExpr(_) => None
+  | TTopData(data_decls) =>
+    Some(
+      List.filter_map(
+        (data_decl: Typedtree.data_declaration) => {
+          let kind: option(ResponseResult.symbol_kind) =
+            switch (data_decl.data_kind) {
+            | TDataRecord(_) => Some(Struct)
+            | TDataVariant(_) => Some(Enum)
+            | TDataAbstract => None
+            };
+          switch (kind) {
+          | Some(kind) =>
+            Some(
+              {
+                name: Ident.name(data_decl.data_id),
+                detail: None,
+                kind,
+                tags: None,
+                range: Utils.loc_to_range(data_decl.data_loc),
+                selection_range: Utils.loc_to_range(data_decl.data_loc),
+                children: None,
+              }: ResponseResult.document_symbol,
+            )
+          | None => None
+          };
+        },
+        data_decls,
+      ),
+    )
+  | TTopLet(_, _, binds) =>
+    Some(
+      List.filter_map(
+        (bind: Typedtree.value_binding) => {
+          switch (bind.vb_pat.pat_desc) {
+          | TPatVar(ident, _) =>
```

Comment:
Shouldn't we at least have `TPatConstraint`? I might argue that we'd want to process every `TPatVar` that appears in a toplevel binding.

Reply:
Doesn't `TPatConstraint` only appear under `pat_extra`?.

As per the others I would generally agree though I was thinking of leaving these for a small follow pr:
* cases like `let [x,y] = [() => void, () => void]` in order to determine that `x` and `y` should have the kinds of functions I would have to start inspecting and breaking down the `exp_type` which seems a little complicated. Alternatively we could just say destructured items are of kind `Variable`.
* I am not exactly sure what the range and selection ranges should be for destructured items.

Reply:
`TPatConstraint` is just when the user writes a type annotation. I don't think looking at the type is too bad, really. If it's an arrow then you know it's a function, and it can be a variable otherwise. As for range, for the big one I'd just do the whole binding, and for the small one I'd do just the name.

Reply:
I think it's fine to start here.